create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java

```diff
@@ -32,28 +38,44 @@ public ScopusHandler(S3Client s3Client) {
     }
 
     @Override
-    public String handleRequest(S3Event event, Context context) {
+    public CreatePublicationRequest handleRequest(S3Event event, Context context) {
         return attempt(() -> readFile(event))
-            .map(this::parseXmlFile)
-            .map(this::getDoi)
-            .orElseThrow(fail -> logErrorAndThrowException(fail.getException()));
+                .map(this::parseXmlFile)
+                .map(this::generateCreatePublicationRequest)
+                .orElseThrow(fail -> logErrorAndThrowException(fail.getException()));
     }
 
-    private String getDoi(DocTp docTp) {
-        var doi= docTp.getMeta().getDoi();
-        logger.info("The publication doi:"+doi);
-        return doi;
+    private CreatePublicationRequest generateCreatePublicationRequest(DocTp docTp) {
+        CreatePublicationRequest createPublicationRequest = new CreatePublicationRequest();
+        createPublicationRequest.setAdditionalIdentifiers(generateAdditionalIds(docTp));
+        return createPublicationRequest;
     }
 
+    private Set<AdditionalIdentifier> generateAdditionalIds(DocTp docTp) {
+        Set<AdditionalIdentifier> additionalIdentifiers = new HashSet<>();
+        docTp
+                .getItem()
+                .getItem()
+                .getBibrecord()
+                .getItemInfo()
+                .getItemidlist()
+                .getItemid()
+                .stream()
+                .filter(itemIdTp -> itemIdTp.getIdtype().equalsIgnoreCase(ScopusConstants.SCOPUS_ITEM_ID_SCP_FIELD_NAME))
+                .forEach(itemIdTp -> additionalIdentifiers.add(new AdditionalIdentifier(ScopusConstants.ADDITIONAL_IDENTIFIERS_SCOPUS_ID_SOURCE_NAME, itemIdTp.getValue())));
+        return additionalIdentifiers;
```

Comment:
I think on the whole, I prefer a more readable style with a single return point:

`

    private Set<AdditionalIdentifier> generateAdditionalIdentifiers(DocTp docTp) {
        return extractItemIdentifiers(docTp)
            .stream()
            .filter(this::isScopusIdentifier)
            .map(this::toAdditionalIdentifier)
            .collect(Collectors.toSet());
    }

    private List<ItemidTp> extractItemIdentifiers(DocTp docTp) {
        return docTp.getItem()
                .getItem()
                .getBibrecord()
                .getItemInfo()
                .getItemidlist()
                .getItemid();
    }

    private boolean isScopusIdentifier(ItemidTp itemIdTp) {
        return itemIdTp.getIdtype().equalsIgnoreCase(ScopusConstants.SCOPUS_ITEM_ID_SCP_FIELD_NAME);
    }

    private AdditionalIdentifier toAdditionalIdentifier(ItemidTp itemIdTp) {
        return new AdditionalIdentifier(ScopusConstants.ADDITIONAL_IDENTIFIERS_SCOPUS_ID_SOURCE_NAME, itemIdTp.getValue());
    }
`

Reply:
Fikset